Menus/BuyMenuGUI.cpp

```diff
@@ -1807,65 +1899,54 @@ void BuyMenuGUI::Update()
             // Events on the Cart List
 
             else if (anEvent.GetControl() == m_pCartList)
-            {
-/*
-                // Somehting was just selected, so update the selection index to the new selected index
-                if(anEvent.GetMsg() == GUIListBox::Select)
-                {
-                    if (m_ListItemIndex != m_pCartList->GetSelectedIndex())
-                        g_GUISound.SelectionChangeSound()->Play(m_pController->GetPlayer());
-                    m_ListItemIndex = m_pCartList->GetSelectedIndex();
-                }
-                // Somehting was clicked upon, therefore should be removed
-                else*/ if(anEvent.GetMsg() == GUIListBox::MouseDown)
-                {
-                    m_pCartList->SetFocus();
-                    m_MenuFocus = ORDER;
-
+            {                
+                if (anEvent.GetMsg() == GUIListBox::MouseUp && !m_IsDragging) {
                     GUIListPanel::Item *pItem = m_pCartList->GetSelected();
                     if (pItem)
                     {
-                        m_ListItemIndex = m_pCartList->GetSelectedIndex();
-                        m_pCartList->ScrollToSelected();
-
-                        if (pItem && pItem->m_pEntity && pItem->m_pEntity->GetClassName() == "AHuman" && g_SettingsMan.SmartBuyMenuNavigationEnabled()) {
-                            int lastItemToDelete = m_pCartList->GetItemList()->size() - 1;
-                            for (int i = m_ListItemIndex + 1; i != m_pCartList->GetItemList()->size(); i++) {
-                                GUIListPanel::Item *cartItem = m_pCartList->GetItem(i);
-                                if (dynamic_cast<const Actor *>(cartItem->m_pEntity)) {
-                                    lastItemToDelete = i - 1;
-                                    break;
+                        if (anEvent.GetData() & GUIListBox::MOUSE_LEFT) {
+                            // We want to allow the user to shift-click to clear the entire cart
+                            // In future it would be nice to add the concept of a modifier key to the controller,
+                            // So we can do this for gamepad inputs too
+                            if (g_UInputMan.FlagShiftState()) {
+                                ClearCartList();
+                                pItem = nullptr;
```

Comment:
Is there any reason not to break out or return or whatever here?
Seems to me that'd be better and cleaner than setting pItem to nullptr here, unless there's something I'm missing by just looking at this little bit of code.

Reply:
We want to handle other events potentiall. If I want to break specifically out of this event, I'd need to subcompartmentalize this into a function, and that seemed like a bunch of effort. I could technically just throw a "continue" in, but that felt a bit ugly because the while loop is so far away and this is such a huge chunk of messy code that it's almost like a goto.